pop_tools/regrid.py

```diff
@@ -0,0 +1,420 @@
+import os
+import warnings
+
+import cf_xarray
+import numpy as np
+import xarray as xr
+
+try:
+    import xesmf as xe
+except ImportError:
+    message = 'Zonal averaging requires xesmf package.\n\n'
+    'Please conda install as follows:\n\n'
+    ' conda install -c conda-forge xesmf>=0.4.0'
+
+    raise ImportError(message)
+
+from tqdm import tqdm
+
+from .grid import _compute_corners, get_grid
+
+
+def gen_corner_calc(ds, cell_corner_lat='ULAT', cell_corner_lon='ULONG'):
+    """
+    Generates corner information and creates single dataset with output
+    """
+
+    cell_corner_lat = ds[cell_corner_lat]
+    cell_corner_lon = ds[cell_corner_lon]
+    # Use the function in pop-tools to get the grid corner information
+    corn_lat, corn_lon = _compute_corners(cell_corner_lat, cell_corner_lon)
+
+    lon_shape, lat_shape = corn_lon[:, :, 0].shape
+    out_shape = (lon_shape + 1, lat_shape + 1)
+
+    # Generate numpy arrays to store destination lats/lons
+    out_lons = np.zeros((out_shape))
+    out_lats = np.zeros((out_shape))
+
+    # Assign the northeast corner information
+    out_lons[1:, 1:] = corn_lon[:, :, 0]
+    out_lats[1:, 1:] = corn_lat[:, :, 0]
+
+    # Assign the northwest corner information
+    out_lons[1:, :-1] = corn_lon[:, :, 1]
+    out_lats[1:, :-1] = corn_lat[:, :, 1]
+
+    # Assign the southwest corner information
+    out_lons[:-1, :-1] = corn_lon[:, :, 2]
+    out_lats[:-1, :-1] = corn_lat[:, :, 2]
+
+    # Assign the southeast corner information
+    out_lons[:-1, 1:] = corn_lon[:, :, 3]
+    out_lats[:-1, 1:] = corn_lat[:, :, 3]
+
+    return out_lats, out_lons
+
+
+def _add_region_mask(ds, grid, cell_center='TLAT'):
+    """
+    Grabs and combines a grid
+    """
+
+    grid = get_grid(grid)
+
+    return xr.merge([ds, grid[['REGION_MASK', 'region_name']]])
+
+
+def _get_default_filename(src_grid, dst_grid, method):
```

Comment:
This is only used once and is not complicated. let's avoid the indirection and stick the code in the `if weight_file is None` block below.